### PR TITLE
Add LittleFS as a new filesystem option

### PR DESCRIPTION
**Pull request by utzig**
This brings in initial LittleFS code from the upstream repo, sha 1863dc7883d82bd6ca79faa164b65341064d1c16, and adds full metadata and glue code to use it with the standard `fs/fs` interface.

Also updated `sys/config` and `apps/slinky` to allow it be used as a backend for config.

All `fs/fs` interfaces were tested on the `frdm-k82f` BSP with an FS test app that I will submit in another PR after some clean up.

PS: I am aware I left some `TODO` or `XXX` in the code, and I'd prefer to leave them there for now.

### CODE REVIEW

File: fs/littlefs/src/mynewt_glue.c [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+
+    file->file = out_file;
+    file->fops = &littlefs_ops;
+    file->lfs = g_lfs;
+    *out_fs_file = (struct fs_file *) file;
+    rc = FS_EOK;
+
+out:
+    if (rc != FS_EOK) {
+        if (file) free(file);
```
Comment (sjanc) [inline fs/littlefs/src/mynewt_glue.c:343]: style issue

File: fs/littlefs/src/mynewt_glue.c [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+        return FS_EFULL;
+    }
+
+    return FS_EOK;
+}
+
+static int
+littlefs_flush(struct fs_file *fs_file)
+{
+    (void)fs_flush;
```
Comment (sjanc) [inline fs/littlefs/src/mynewt_glue.c:520]: fs_flush()   ?

File: fs/littlefs/src/mynewt_glue.c [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+littlefs_unlink(const char *path)
+{
+    int rc;
+
+    if (!path) {
+        return FS_EINVAL;
+    }
+
+    littlefs_lock();
+    rc= lfs_remove(g_lfs, path);
```
Comment (sjanc) [inline fs/littlefs/src/mynewt_glue.c:534]: style issue

File: fs/littlefs/src/mynewt_glue.c [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+    dir->dir = out_dir;
+    dir->cur_dirent = NULL;
+    dir->fops = &littlefs_ops;
+    dir->lfs = g_lfs;
+    *out_fs_dir = (struct fs_dir *)dir;
+    rc = FS_EOK;
+
+out:
+    if (rc != FS_EOK) {
+        if (dir) free(dir);
```
Comment (sjanc) [inline fs/littlefs/src/mynewt_glue.c:614]: style
also doesn't free() already check for null pointers?

File: fs/littlefs/src/mynewt_glue.c [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+
+    dirent = ldir->cur_dirent;
+    dir = ldir->dir;
+    lfs = ldir->lfs;
+
+    littlefs_lock();
+    rc = lfs_dir_read(lfs, dir, &dirent->info);
+    littlefs_unlock();
+    if (rc < 0) {
+        return littlefs_to_vfs_error(rc);
```
Comment (sjanc) [inline fs/littlefs/src/mynewt_glue.c:651]: isn't dirent leaked here?

File: fs/littlefs/src/mynewt_glue.c [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+    }
+
+    if (rc == 0) {
+        return FS_ENOENT;
```
Comment (sjanc) [inline fs/littlefs/src/mynewt_glue.c:655]: ditto

File: fs/littlefs/src/mynewt_glue.c [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+    }
+
+    /*
+     * This doesn't seem to be needed because lfs_mount initializes
+     * all fields, but just to stay on the safe side...
+     */
+    memset(g_lfs, 0, sizeof(lfs_t));
+
+    rc = flash_area_open(MYNEWT_VAL(LITTLEFS_FLASH_AREA), &fa);
+    if (rc) {
```
Comment (sjanc) [inline fs/littlefs/src/mynewt_glue.c:757]: g_lfs leaked here

File: fs/littlefs/src/mynewt_glue.c [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+    if (g_lfs && g_lfs_alloc_done) {
+        return FS_EOK;
+    }
+
+    rc = os_mutex_init(&littlefs_mutex);
+    if (rc != 0) {
+        return FS_EOS;
+    }
+
+    g_lfs = malloc(sizeof(lfs_t));
```
Comment (sjanc) [inline fs/littlefs/src/mynewt_glue.c:745]: g_lfs will leak on subsequent call (if previous one failed on flash_area_open())

**Review by sjanc — COMMENTED** [review]
some comments inline

**Comment by sjanc** [comment]
also, I wonder if when this is merge should we deprecate nffs (and remove it eventually)

**Comment by utzig** [comment]
> also, I wonder if when this is merge should we deprecate nffs (and remove it eventually)

I would not go that far at least for this next release, and thanks for the review btw, I will fix the issues soon.

File: fs/littlefs/syscfg.yml [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    LITTLEFS_FLASH_AREA:
+        description: 'Flash area to use for littlefs.'
+        type: flash_owner
+        value:
+        restrictions:
+            - $notnull
+
+    LITTLEFS_DETECT_FAIL:
+        description: 'Controls behaviour when encountering corrupt littlefs area.'
+        value: 'LITTLEFS_DETECT_FAIL_FORMAT'
```
Comment (andrzej-kaczmarek) [inline fs/littlefs/syscfg.yml:29]: I do not like using preprocessor definitions as syscfg values as effectively this injects raw source code via syscfg setting. This should better be 'choice' insead.

File: fs/littlefs/src/mynewt_glue.c (hunk 4 above)
Comment (utzig) [reply fs/littlefs/src/mynewt_glue.c]: Fixed.

File: fs/littlefs/src/mynewt_glue.c (hunk 5 above)
Comment (utzig) [reply fs/littlefs/src/mynewt_glue.c:652]: Yep, fixed.

File: fs/littlefs/src/mynewt_glue.c (hunk 8 above)
Comment (utzig) [reply fs/littlefs/src/mynewt_glue.c:747]: Fixed.

File: fs/littlefs/syscfg.yml (hunk 9 above)
Comment (utzig) [reply fs/littlefs/syscfg.yml]: Done.

**Comment by utzig** [comment]
@sjanc @andrzej-kaczmarek I believe all issues were addressed, just need to merge #2526 go get it passing (as well as all other PRs opened today!).

**Review by sjanc — APPROVED** [review]

**Review by andrzej-kaczmarek — APPROVED** [review]

**Pull request merged**